cmd/gcp-controller-manager/node_csr_approver.go

```diff
@@ -985,16 +988,25 @@ func shouldDeleteNode(ctx *controllerContext, node *v1.Node, getInstance func(*c
 }
 
 func deleteAllPodsBoundToNode(ctx *controllerContext, nodeName string) error {
-	errs := []error{}
 	deletedPods := []string{}
 
-	podList, err := ctx.client.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{FieldSelector: "spec.nodeName=" + nodeName})
+	podList, err := ctx.client.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{FieldSelector: "spec.nodeName=" + nodeName, ResourceVersion: "0"})
 	if err != nil {
-		errs = append(errs, err)
+		return fmt.Errorf("failed to list pods on node: %v: %v", nodeName, err)
 	}
 
+	errs := []error{}
 	for _, pod := range podList.Items {
-		err := ctx.client.CoreV1().Pods(pod.Namespace).Delete(context.TODO(), pod.Name, *metav1.NewDeleteOptions(0))
+		pod := pod
+
+		condition := corev1apply.PodCondition().
+			WithType(v1.DisruptionTarget).
+			WithStatus(v1.ConditionTrue).
+			WithReason("DeletionByGCPControllerManager").
+			WithMessage("GCPControllerManager: node no longer exists").
+			WithLastTransitionTime(metav1.Now())
+
+		err := markFailedAndDeletePodWithCondition(ctx, &pod, condition)
```

Comment:
if it's already Completed or Failed, do we need to actually delete the Pod?

Reply:
We don't need to delete the pod, however, we want to follow the same logic as pod-gc-controller which will delete the pod once the node is gone - https://github.com/kubernetes/kubernetes/blob/c9ed04762f94a319d7b1fb718dc345491a32bea6/pkg/controller/podgc/gc_controller.go#L363

Reply:
If GCPControllerManager weren't to delete the pod what would delete it? IIUC the Job tracking finalizer will still be removed by the job_controller, even though the node is replaced?

Reply:
If the pod is already Completed or Failed, we shouldn't add the condition either. 

@mimowo isn't this a bug in upstream?

Reply:
After a quick code inspection, I didn't see that those pods are filtered. We should still delete them, but adding the condition is not accurate.

Reply:
I think it is ok to add the condition, it is anyway possible that a pair of controllers would like to set the condition and only one wins, thus we add the condition unconditionally in general.

Reply:
I see, I guess we could then have it as a bug in upstream, but isn't it a similar problem we considered before - should we add the condition is one is already added by another controller. We decided that we still add it because it does not make much of a difference which one is the last. If we add this if it is still possible that another controller will change the phase or add the condition in the meanwhile.

Reply:
No, it's not the same.

In this case, the pod already finished by its own merit (or because some other controller already terminated it). In either scenario, the Pod GC shouldn't add the condition.

Reply:
In case of PodGC we only invoke the function with `condition!=nil` for orphaned pods, which are typically left in the Running state, because there is no kubelet to terminate them, but indeed it may happen that there is an orphaned Failed pod. Should I open an issue for this?

Reply:
The thing is that Job pods (and plain pods) are not deleted when they finish (until the regular pod gc catches them).

Reply:
Oh, but as long as they are not orphaned, PodGC does not attempt to add the condition at all. They will be handled typically by gcTerminated: https://github.com/kubernetes/kubernetes/blob/4802d7bb62c2623be8e4f940f6b5c1fcddd6c744/pkg/controller/podgc/gc_controller.go#L182

Reply:
I meant this place: https://github.com/kubernetes/kubernetes/blob/4802d7bb62c2623be8e4f940f6b5c1fcddd6c744/pkg/controller/podgc/gc_controller.go#L215

Reply:
right, but they could still get orphaned some time after they finish.

Reply:
for example due to scale down, ok I guess this convinces me, will open an Issue.

Reply:
Thanks for discussion, updated the logic to only set condition if the phase is not terminal (i.e. not Succeeded or Failed).

Reply:
Opened the PR for PodGC: https://github.com/kubernetes/kubernetes/pull/115056